### PR TITLE
IGNITE-21330 Sql. Support index scan for OR operator with dynamic parameters

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-21330

Calcite can compose OR operator with literals to SEARCH/SARG, and SEARCH/SARG can be used for index scans. But we can't do this for OR operator with dynamic parameters. 

For example expression a IN (1, 2, 3) can be converted to a = 1 OR a = 2 OR a = 3 and after that can be converted to SEARCH(a, SARG(1, 2, 3)), but expression a IN (?, ?, ?) can be converted only to a = ? OR a = ? OR a = ? and can't be used for index scan.

To fix this issue we can create ranges from dynamic parameters during planning, sort and remove intersections from these ranges in runtime.

Note: this patch is a port of Apache Ignite 2 PR https://github.com/apache/ignite/pull/11126

The implementation consists of two parts:

* Assembly of search boundaries for `OR` operator during planning (see `RexUtils`)
* Compare and merge ranges in runtime (see `ExpressionFactoryImpl`)

### CODE REVIEW

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSecondaryIndexMultiRangeScanTest.java [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+                .check();
+    }
+
+    /** Test correct index ordering without additional sorting. */
+    @ParameterizedTest(name = "dynamicParams={0}")
+    @ValueSource(booleans = {true, false})
+    public void testOrdering(boolean useDynamicParameters) {
+        assertQuery(useDynamicParameters, "SELECT * FROM test_asc WHERE c1 IN (%s, %s) AND c2 IN (%s, %s) ORDER BY c1, c2, c3",
+                3, 2, "3", "2")
+                .matches(CoreMatchers.not(QueryChecker.containsSubPlan("Sort"))) // Don't require additional sorting.
```
Comment (lowka) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSecondaryIndexMultiRangeScanTest.java:189]: These tests reply on index usage, why not to use USE_INDEX hint directly?

Comment (lowka) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSecondaryIndexMultiRangeScanTest.java:189]: I meant `FORCE_INDEX`.

Comment (xtern) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSecondaryIndexMultiRangeScanTest.java:192]: Done

**Review by lowka — APPROVED** [review]

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSecondaryIndexRandomConditionsTest.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        for (int i = 0; i < seeds.length; i++) {
+            seeds[i] = rnd.nextLong();
+        }
+    }
+
+    @ParameterizedTest(name = "order direction = {0}")
+    @ValueSource(strings = {"ASC", "DESC"})
+    @WithSystemProperty(key = "IMPLICIT_PK_ENABLED", value = "true")
+    public void testRandomConditions(String direction) {
+        sql("create table test0(a int, b int, c int) with primary_zone='IN_MEM_ZONE'");
```
Comment (lowka) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSecondaryIndexRandomConditionsTest.java:66]: Maybe we can move setup into before all method?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/exp/ExpressionFactoryImplTest.java [hunk 3]
```diff
@@ -311,6 +312,232 @@ void multiBoundConditionAreOrderedCorrectly() {
...
+                            new ExactBounds(condition, intValue1),
+                            new RangeBounds(condition, null, intValue1, true, false)
+                    ))
+            );
+
+            Comparator<Object[]> comparator = expFactory.comparator(RelCollations.of(
+                    new RelFieldCollation(0, Direction.ASCENDING)));
+
+            RangeIterable<Object[]> ranges = expFactory.ranges(boundsList, rowType, comparator);
+            List<TestRange> list = new ArrayList<>();
```
Comment (lowka) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/exp/ExpressionFactoryImplTest.java:344]: I think collecting a list of ranges from a range iterable could be extracted into a separate method to improve readability of these test cases.

**Review by korlov42 — APPROVED** [review]

**Pull request merged**